### ISSUE TITLE
Support for encrypted unload?

### ISSUE THREAD

**Issue by gapar2** [databricks/spark-redshift#189]
Redshift has the option of doing an [encrypted unload operation](http://docs.aws.amazon.com/redshift/latest/dg/t_unloading_encrypted_files.html). Is that something that is on your roadmap, and if not, do you have a rough idea of how much work would be involved?

Edit: I've been looking into the code. I don't have much experience in scala, but it seems [Parameters.scala](https://github.com/databricks/spark-redshift/blob/master/src/main/scala/com/databricks/spark/redshift/Parameters.scala) would need to change to parse/validate the encryption parameter. Also, [RedshiftRelation](https://github.com/databricks/spark-redshift/blob/d97ffb0b607076a31693e93f802ff410767afbf6/src/main/scala/com/databricks/spark/redshift/RedshiftRelation.scala) would need to know how to buildUnloadStmt with the encrypted parameter. Then there'd be a step missing before Conversions.scala which would actually decrypt the data before it can be converted, is this correct? 

Overall, is this something that would be easily doable, or would it require significant refactoring?

**Comment by JoshRosen** [comment]
It looks like you have a pretty good summary of the changes that would be necessary.

> Then there'd be a step missing before Conversions.scala which would actually decrypt the data before it can be converted, is this correct?

Yep, I think this would be the tricky part. There are [instructions](https://aws.amazon.com/articles/2850096021478074) on using this type of encryption directly with the AWS S3 SDK (see also https://aws.amazon.com/blogs/aws/client-side-data-encryption-using-the-aws-sdk-for-java/) but I don't think that's going to work when we're interacting with S3 via the Hadoop S3A or S3N filesystems.

I think this would work when using EMRFS to access S3, though: https://aws.amazon.com/about-aws/whats-new/2015/03/amazon-emr-support-for-amazon-s3-client-side-encryption-now-available/

So, in summary, I think this would take a bit of work.

**Comment by gapar2** [comment]
Thank you for the quick reply. I'll find a workaround for now. Great job with the package BTW, this greatly simplifies interaction between redshift and spark.

**Comment by JoshRosen** [comment]
Actually, I did think of one good workaround: given that our read path starts with a list of S3 file paths (read from a manifest (which isn't encrypted)), you could modify the actual S3 reading path to use the Amazon client rather than going through the Hadoop APIs. Here's the relevant part of the code: https://github.com/databricks/spark-redshift/blob/cba5eee1ab79ae8f0fa9e668373a54d2b5babf6b/src/main/scala/com/databricks/spark/redshift/RedshiftRelation.scala#L149

What you could do is to do something like `sc.parallelize(filesToRead, filesToRead.length)` and then use a `mapPartitions` call to create an AmazonS3EncryptionClient, read the file, decrypt the bytes, and pass them to the input format code (which you could extract and decouple from the Hadoop interfaces). Presumably the encrypted files aren't splittable by Hadoop anyways, so the fact that you'd be losing any benefits of FileInputFormat splitting (not sure if we're actually taking advantage of that, though).

This is a little complicated but I'm pretty sure that it would work.

**Comment by jiteshsoni** [comment]
Hi Josh,

I used S3N File system to write back data to S3. With the below configuration, I am able to use Spark Redshift encryption thus ensuring Encryption at Rest.
    sc._jsc.hadoopConfiguration().set("fs.s3n.awsAccessKeyId", aws_access_key_id)
    sc._jsc.hadoopConfiguration().set(
        "fs.s3n.awsSecretAccessKey", aws_secret_access_key)
    sc._jsc.hadoopConfiguration().set("fs.s3n.server-side-encryption-algorithm", "AES256")

**Comment by jiteshsoni** [comment]
Hi Josh,

Could you help me understand on how to ensure encryption in transit ?

**Comment by JoshRosen** [comment]
I'm in the process of writing a README update for this. Here's my notes on the current encryption support:

- **Securing JDBC**: The Redshift and Postgres JDBC drivers both support SSL. To enable SSL support,
    first configure Java to add the required certificates by following the
    [_Using SSL and Server Certificates in Java_](http://docs.aws.amazon.com/redshift/latest/mgmt/connecting-ssl-support.html#connecting-ssl-support-java)
    instructions in the Redshift documentation. Then, follow the instructions in
    [_JDBC Driver Configuration Options_](http://docs.aws.amazon.com/redshift/latest/mgmt/configure-jdbc-options.html) to add the appropriate SSL options
    to the JDBC `url` that you use with this library.
    
- **Encrypting `UNLOAD` data stored in S3 (data stored when reading from Redshift)**: According to the Redshift documentation
    on [_Unloading Data to S3_](http://docs.aws.amazon.com/redshift/latest/dg/t_Unloading_tables.html), 
    "UNLOAD automatically encrypts data files using Amazon S3 server-side encryption (SSE-S3)."
    
    Redshift also supports client-side encryption with a custom key
    (see: [_Unloading Encrypted Data Files_](http://docs.aws.amazon.com/redshift/latest/dg/t_unloading_encrypted_files.html))
    but this library currently lacks the capability to specify the required symmetric key.
    
- **Encrypting `COPY` data stored in S3 (data stored when writing to Redshift)**:
    According to the Redshift documentation on 
    [_Loading Encrypted Data Files from Amazon S3_](http://docs.aws.amazon.com/redshift/latest/dg/c_loading-encrypted-files.html):
    
    > You can use the COPY command to load data files that were uploaded to Amazon S3 using
    > server-side encryption with AWS-managed encryption keys (SSE-S3 or SSE-KMS), client-side
    > encryption, or both. COPY does not support Amazon S3 server-side encryption with a customer-supplied key (SSE-C)
    
    To use this capability, you should configure your Hadoop S3 FileSystem to use encryption by 
    setting the appropriate configuration properties (which will vary depending on whether you
    are using `s3a`, `s3n`, EMRFS, etc.).
    Note that the `MANIFEST` file (a list of all files written) will not be encrypted.
